modules/misc/listenbrainz.c

```diff
@@ -0,0 +1,585 @@
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <assert.h>
+#include <time.h>
+
+#include <vlc_common.h>
+#include <vlc_plugin.h>
+#include <vlc_interface.h>
+#include <vlc_input_item.h>
+#include <vlc_dialog.h>
+#include <vlc_meta.h>
+#include <vlc_memstream.h>
+#include <vlc_stream.h>
+#include <vlc_url.h>
+#include <vlc_tls.h>
+#include <vlc_player.h>
+#include <vlc_playlist.h>
+#include <vlc_vector.h>
+
+typedef struct listen_t
+{
+    char *psz_artist;
+    char *psz_title;
+    char *psz_album;
+    char *psz_track_number;
+    int i_length;
+    char *psz_musicbrainz_id;
+    time_t date;
+} listen_t;
+
+typedef struct VLC_VECTOR (listen_t) vlc_vector_listen_t;
+
+struct intf_sys_t
+{
+    vlc_vector_listen_t queue;
+
+    vlc_playlist_t *playlist;
+    struct vlc_playlist_listener_id *playlist_listener;
+    struct vlc_player_listener_id *player_listener;
+    struct vlc_player_timer_id *timer_listener;
+
+    vlc_mutex_t lock;
+    vlc_cond_t wait;                // song to submit event
+    vlc_thread_t thread;            // thread to submit song
+
+    vlc_url_t p_submit_url;         // where to submit data
+    char *psz_user_token;           // authentication token
+
+    listen_t p_current_song;
+    bool b_meta_read;               // check if song metadata is already read
+
+    vlc_tick_t time_played;
+};
+
+static int Open (vlc_object_t *);
+
+static void Close (vlc_object_t *);
+
+static void *Run (void *);
+
+#define USER_TOKEN_TEXT      N_("User token")
+#define USER_TOKEN_LONGTEXT  N_("The user token of your ListenBrainz account")
+#define URL_TEXT             N_("Submission URL")
+#define URL_LONGTEXT         N_("The URL set for an alternative ListenBrainz instance")
+
+/****************************************************************************
+ * Module descriptor
+ ****************************************************************************/
+
+vlc_module_begin ()
+    set_category(CAT_INTERFACE)
+    set_subcategory(SUBCAT_INTERFACE_CONTROL)
+    set_shortname (N_ ("ListenBrainz"))
+    set_description (N_ ("Submit listens to ListenBrainz"))
+    add_string("listenbrainz_user_token", "", USER_TOKEN_TEXT, USER_TOKEN_LONGTEXT, false)
+    add_string("listenbrainz_submission_url", "api.listenbrainz.org", URL_TEXT, URL_LONGTEXT, false)
+    set_capability("interface", 0)
+    set_callbacks(Open, Close)
+vlc_module_end ()
+
+static void DeleteSong (listen_t *p_song)
+{
+    FREENULL (p_song->psz_artist);
+    FREENULL (p_song->psz_album);
+    FREENULL (p_song->psz_title);
+    FREENULL (p_song->psz_musicbrainz_id);
```

Comment:
Maybe use something like this to stay open for including more MBIDs later?
```suggestion
    FREENULL (p_song->psz_musicbrainz_recording_id);
```